### PR TITLE
mds/client: switch to use ceph_assert() instead of assert()

### PR DESCRIPTION
**Pull request by lxbsz**
If the -DNDEBUG was specified when build the code, the assert()
will do nothing.

Fixes: https://tracker.ceph.com/issues/51956
Signed-off-by: Xiubo Li <xiubli@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by tchaikov** [comment]
how come the behavior depends on the difference between `assert()` and `ceph_assert()` ?

**Comment by lxbsz** [comment]
> how come the behavior depends on the difference between `assert()` and `ceph_assert()` ?
For the Shaman builds, for example in https://jenkins.ceph.com/job/ceph-dev-new-build/ARCH=arm64,AVAILABLE_ARCH=arm64,AVAILABLE_DIST=centos8,DIST=centos8,MACHINE_SIZE=gigantic/56460//consoleFull

```
/usr/bin/cc -DDYNAMIC_ANNOTATIONS_ENABLED=1 -DNDEBUG -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -fasynchronous-unwind-tables -fstack-clash-protection -D_GNU_SOURCE -fPIC -fwrapv -iquote/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/arm64/AVAILABLE_ARCH/arm64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/17.0.0-6374-g1762da2b/rpm/el8/BUILD/ceph-17.0.0-6374-g1762da2b/src/include -w -Dvoid0=dead_function(void) -D__Pyx_check_single_interpreter(ARG)=ARG ## 0 -fPIC -I/usr/include/python3.6m -I/usr/include/python3.6m -c /home/jenkins-build/build/workspace/ceph-dev-new-
```

We can see the `-DNODEBUG` is specified, then in `/usr/include/assert.h`:

```
 48 #ifdef  NDEBUG
 49 
 50 # define assert(expr)           (__ASSERT_VOID_CAST (0))                                                                                           
 51 
 52 /* void assert_perror (int errnum);
 53 
 54    If NDEBUG is defined, do nothing.  If not, and ERRNUM is not zero, print an
 55    error message with the error text for ERRNUM and abort.
 56    (This is a GNU extension.) */
 57 
 58 # ifdef __USE_GNU
 59 #  define assert_perror(errnum) (__ASSERT_VOID_CAST (0))
 60 # endif
 61 
 62 #else /* Not NDEBUG.  */
...
```

Then the `assert()` will do nothing.

So for the PR#41969, because in https://github.com/ceph/ceph/blob/master/src/mds/Migrator.cc#L1942 is using the `assert()`, the local test works well, but not for the teuthology test which is using the ceph binaries built by the Shaman.

**Comment by tchaikov** [comment]
i meant why the *behavior* is different. as both statements should not have any side effect. 
in the case of
```c++
assert (g_conf()->mds_kill_export_at != 10);
```
neither `ceph_assert()` nor `assert()` has any side effect.

**Comment by lxbsz** [comment]
> i meant why the _behavior_ is different. as both statements should not have any side effect.
> in the case of
> 
> ```c++
> assert (g_conf()->mds_kill_export_at != 10);
> ```
> 
> neither `ceph_assert()` nor `assert()` has any side effect.

How about something like:

```
diff --git a/src/mds/Migrator.cc b/src/mds/Migrator.cc
index 14d9799c109..e3f015279c5 100644
--- a/src/mds/Migrator.cc
+++ b/src/mds/Migrator.cc
@@ -1939,7 +1939,9 @@ void Migrator::handle_export_ack(const cref_t<MExportDirAck> &m)
   // log export completion, then finish (unfreeze, trigger finish context, etc.)
   mds->mdlog->submit_entry(le, new C_MDS_ExportFinishLogged(this, dir));
   mds->mdlog->flush();
-  ceph_assert(g_conf()->mds_kill_export_at != 10);
+  if (g_conf()->mds_kill_export_at == 10) {
+    ceph_abort();
+  }
 }
```

**Comment by lxbsz** [comment]
The first two commits still make sense, we need to make sure that odd issue, which in different builds it have different behaviors, won't happen in future.

File: src/mds/Migrator.cc [hunk 1]
```diff
@@ -1407,7 +1407,7 @@ void Migrator::export_frozen(CDir *dir, uint64_t tid)
   // send.
   it->second.state = EXPORT_PREPPING;
   mds->send_message_mds(prep, it->second.peer);
-  assert (g_conf()->mds_kill_export_at != 4);
+  ceph_assert(g_conf()->mds_kill_export_at != 4);
```
Comment (tchaikov) [inline src/mds/Migrator.cc]: oic. `mds_kill_export_at` is used for fault injection. so it should not be optimized out.

**Comment by lxbsz** [comment]
jenkins test make check arm64

**Comment by lxbsz** [comment]
jenkins test make check arm64

**Review by batrick — APPROVED** [review]
In the future, you can combine the commits with title "client,mds: ...".

**Comment by lxbsz** [comment]
> In the future, you can combine the commits with title "client,mds: ...".

Sure. Thanks.

**Comment by batrick** [comment]
https://tracker.ceph.com/projects/cephfs/wiki/Master#2021-September-10

**Pull request merged**